src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+package scalaz.stream.async
+
+import scalaz.stream.Process
+import scalaz.stream.Process.{Halt, Step, Emit, Await, Cont}
+import scalaz.{Trampoline, \/}
+import scalaz.\/._
+import scalaz.concurrent.{Strategy, Actor, Future, Task}
+import scalaz.stream.Cause
+import scalaz.stream.Cause.{Kill, EarlyCause}
+import scalaz.stream.Util._
+
+/**
+ * Asynchronous driver for Process, where `F` is specialized to TASK 
+ */
+object TaskDriver {
+
+  /**
+   * Runs the supplied task with possible chance to interrupt it.
+   * Task is interrupted when resulting callback is applied.
+   *
+   * Whenever task is interrupted, following is guaranteed:
+   * - if, there was asynchronous code of task in process, the `cb` will get invoked immediatelly with `Task.TaskInterrupted` exception
+   * - if the task is complete, then interruption is no-op
+   * - if the task is running `pure` code, that code is completed and `cb` is invoked with result of that code
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
```

Comment:
> I think the failure of the later is correct, hence the inner part was never consulted, because the await was interrupted before you entered in rcv here. I think this is perfectly consistent, as the interrupt behaves same way as if the task will fail in that case you as well won't consulte any inner cleanup.

Except the task isn't failing, it's completing!  I'll consolidate this discussion with the one on my specific example below, as they are in fact the same test case.

> If you perhaps have any other cases can you maybe at them to this PR so we can assure they will execute as well? Thanks a lot!

It's mostly the test case with the "inner finalizer invocation count" failure that is interesting.

> I still believe that this implementation guarantees exactly once semantic. Can you please write test case in which this does not hold? The onComplete here is executed for both scenarios when task completes with value and when task does not complete with value, but only once.

The test case which fails from my branch is exactly the test case in which exactly once doesn't hold.  :-)  There is an `onComplete` which is not invoked.  As I said, exactly once isn't really possible, but we *can* provide much stronger guarantees than what this branch provides, which is precisely what my implementation does.

> I don't think so the tryCompleteTask has same semantics as runAsyncInterruptibly but maybe I am wrong here. What the implementation does it puts the decision on the actor, and thus we guarantee here that we interrupt only asynchronous code and exit from the loop after very first asynchronous code being interrupted. In other words We won't interrupt any pure code, which I think is undesirable.

Remember that `delay` isn't necessarily pure.  In fact, idiomatically, it most often is not.

`tryCompleteTask` really does have exactly the same semantics as `runAsyncInterruptibly` except that it will not interrupt at bind points on tasks constructed with `delay`, which as I said, I believe is a misfeature.  The really critical point is that `tryCompleteTask` completely discards values upon completion if the task has been previously interrupted, which leaks resources.  This is the real reason why `runAsyncInterruptibly` is inappropriate for this use case.  If we're willing to accept resource leaks (which was incidentally the original objection to my PR!), then we may as well use the standard interpreter.

>  Additionally by having actor in loop I think is a bit easier to reason about what gets executed before and after, but maybe its sort of taste than anything else.
>
> I am not an actor - lover, but I found them useful for these type of problem solving.

I'm more interested in the factoring of separate semantic cases.  For example, how does your code handle the case where we interrupted the task, but it had already completed and passed along to the next process iteration, and thus we have to interrupt *that* iteration?  I know it handles it, but can you point to *one* spot in your code where that case is identified and resolved?  As far as I can tell, the logic for this case is quite spread out, which makes it tricky to reason about.

> As you mention, and as I explained above I think it is undesirable to interrupt the Task.delay or Bind code. I don't see any real use case for it, and for me this just brings another unnecessary level of nondeterminism (e.g. why you would need to nondeterministically interrupt pure code?). You mention you see it as significant limitation, can you please explain this a bit more? If we really need it, than yeah, we can use standard Task#runAsyncInterruptibly perhaps.

It is extremely standard to encapsulate impure code in `Task.delay`.  That's how nearly all `Task` code (that I've seen, including scalaz-stream itself!) is implemented.  `Task.apply` is a somewhat less common constructor.  Your code will *only* interrupt `Task.apply` and binds from `Task.apply`.  Interrupting `Task.delay` is extremely important for the same reason that interrupting `Task.apply` is important.

Remember that the interrupt doesn't actually interrupt the evaluation of an atomic task block!  It just suppresses the value from passing along to the next step in the chain.  This is in fact *precisely* the semantic we want with `await`: interrupt the computation as early as possible to free up CPU/memory resources and *hopefully* prevent side-effects from getting triggered in the first place.